File: translator.py
# Author: Md. Shifat Hasan (2107067)

import sys

OpCodes = ['noop', 'set', 'clr', 'not', 'neg', 'and', 'or', 'xor', 'add', 'sub', 'mul', 'div', 'shl',
           'shr', 'cmp', 'jmp', 'jmpz', 'jmpl', 'jmpg', 'ina', 'inx', 'outa', 'outx', 'showf', 'show', 'hlt']

UnaryOpCodes = ['noop', 'not', 'neg', 'shl', 'shr', 'showf', 'hlt']

BinaryOpCodes = ['set', 'clr', 'and', 'or', 'xor', 'add', 'sub', 'mul', 'div',
                 'cmp', 'jmp', 'jmpz', 'jmpl', 'jmpg', 'ina', 'inx', 'outa', 'outx', 'show']


def opcode_hex_value(opcode: str):
    id = OpCodes.index(opcode)
    id = hex(id)[2:].zfill(2)
    return id


def content_from_lines(lines: list) -> str:
    return '\n'.join(lines)


def preprocess(source_file: str) -> str:
    with open(source_file, 'r') as file:
        lines = file.readlines()

    # Process the lines in a single loop
    processed_lines = []
    for line in lines:
        line = line.lower().replace(':', ':\n').strip()  # Lowercase, replace, and strip in one step
        line = line.split('!')[0].strip()  # Remove comments and trailing whitespace
        if line:  # Only add non-empty lines
            processed_lines.append(line)

    return content_from_lines(processed_lines)


def vars_and_instructions(content: str) -> tuple:
    lines = content.splitlines()
    vars = [line for line in lines if line.startswith('$')]
    labels = {}
    instructions = []

    for line in lines:
        if not line:
            continue
        if line.startswith('@'):
            label = ''.join(line.split())[:-1]
            if label not in labels:
                labels[label] = len(instructions)
        elif not line.startswith('$'):
            instructions.append(line)

    return vars, labels, instructions


def validate_instructions(instructions: list, vars: list, labels: list) -> bool:
    hex_chars = set('0123456789abcdefABCDEF')
    binary_chars = set('01')

    def is_valid_address(address: str, opcode: str) -> bool:
        if opcode in ['set', 'clr']:
            valid_flag_numbers = {
                '0', '1', '2', '3',
                '#0', '#1', '#2', '#3',
                '#d0', '#d1', '#d2', '#d3',
                '#h0', '#h1', '#h2', '#h3'
            }
            return address in valid_flag_numbers

        if address in vars or address in labels:
            return True

        if address[0] == '#':
            if address[1] == 'd':
                return address[2:].isdigit()
            elif address[1] == 'h':
                return all(c in hex_chars for c in address[2:])
            elif address[1] == 'b':
                return all(c in binary_chars for c in address[2:])
            else:
                return address[1:].isdigit()

        return address.isdigit()

    for instruction in instructions:
        instruction_parts = instruction.split(' ')
        if not instruction_parts:
            print(f"[ ERROR : Invalid Instruction ] --> '{instruction}'")
            return False

        opcode = instruction_parts[0]
        if opcode not in OpCodes:
            print(f"[ ERROR : Invalid Instruction ] --> '{instruction}'")
            return False

        if (opcode in UnaryOpCodes and len(instruction_parts) > 1) or \
           (opcode in BinaryOpCodes and len(instruction_parts) < 2):
            print(f"[ ERROR : Invalid Instruction ] --> '{instruction}'")
            return False

        address = instruction_parts[1] if len(instruction_parts) > 1 else '0'

        if not is_valid_address(address, opcode):
            print(f"[ ERROR : Invalid Instruction ] --> '{instruction}'")
            return False

    return True


def translate(source_file: str, output_file: str):
    preprocessed_data = preprocess(source_file)  # Stage: 1

    vars, labels, instructions = vars_and_instructions(preprocessed_data)  # Stage: 2

    labels_list = list(labels.keys())
    vars_dict = {}
    vars_list = []

    # Processing variables
    for var in vars:
        var_name, var_value = (var.split(' = ') + ['0'])[:2]

        if var_value.startswith('#'):
            base = {'d': 10, 'h': 16, 'b': 2}.get(var_value[1], 10)
            var_value = int(var_value[2:], base)
        else:
            var_value = int(var_value, 10)

        vars_dict[var_name] = var_value
        if var_name not in vars_list:
            vars_list.append(var_name)

    # Ensure 'hlt' instruction is present
    if not instructions or instructions[-1] != 'hlt':
        instructions.append('hlt')

    if not validate_instructions(instructions, vars_list, labels_list):  # Stage: 3
        print("Translation Failed!")
        return

    lines_with_resolved_addresses = []
    var_init = len(instructions)

    # Resolving addresses
    for instruction in instructions:
        parts = instruction.split()
        opcode = parts[0]
        var = parts[1] if len(parts) > 1 else '0'

        if var.startswith('#'):
            base = {'d': 10, 'h': 16, 'b': 2}.get(var[1], 10)
            var = int(var[2:], base)

        if var in vars_list:
            var = var_init + vars_list.index(var)
        elif var in labels_list:
            var = labels[var]

        address = hex(int(var))[2:].zfill(5)
        opcode = opcode_hex_value(opcode)
        lines_with_resolved_addresses.append(f"{opcode}{address}")

    # Adding variable values to the end
    for var in vars_list:
        lines_with_resolved_addresses.append(f"{hex(vars_dict[var])[2:].zfill(7)}")

    # Building RAM content
    ram_content = 'v3.0 hex words addressed\n00000: ' + ' '.join(lines_with_resolved_addresses)

    with open(output_file, 'w') as RAM:
        RAM.write(ram_content)

    print(f"Translation Successfully Done.\nOutput file: '{output_file}'")


if __name__ == '__main__':
    print()
    if len(sys.argv) != 3:
        print("Usage: python translator.py <source_file> <output_file>")
    else:
        source_file = sys.argv[1]
        output_file = sys.argv[2]
        translate(source_file, output_file)
    print()

File: translator.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var (
	OpCodes = []string{"noop", "set", "clr", "not", "neg", "and", "or", "xor", "add", "sub", "mul", "div", "shl",
		"shr", "cmp", "jmp", "jmpz", "jmpl", "jmpg", "ina", "inx", "outa", "outx", "showf", "show", "hlt"}

	UnaryOpCodes = []string{"noop", "not", "neg", "shl", "shr", "showf", "hlt"}

	BinaryOpCodes = []string{"set", "clr", "and", "or", "xor", "add", "sub", "mul", "div",
		"cmp", "jmp", "jmpz", "jmpl", "jmpg", "ina", "inx", "outa", "outx", "show"}
)

func opcodeHexValue(opcode string) string {
	id := indexOf(OpCodes, opcode)
	return fmt.Sprintf("%02x", id)
}

func contentFromLines(lines []string) string {
	return strings.Join(lines, "\n")
}

func preprocess(sourceFile string) string {
	file, err := os.Open(sourceFile)
	if err != nil {
		fmt.Println("Error opening file:", err)
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var processedLines []string
	for scanner.Scan() {
		line := strings.ToLower(scanner.Text())
		line = strings.Replace(line, ":", ":\n", -1)
		line = strings.Split(line, "!")[0]
		line = strings.TrimSpace(line)
		if line != "" {
			processedLines = append(processedLines, line)
		}
	}
	return contentFromLines(processedLines)
}

func varsAndInstructions(content string) ([]string, map[string]int, []string) {
	lines := strings.Split(content, "\n")
	vars := []string{}
	labels := make(map[string]int)
	instructions := []string{}

	for _, line := range lines {
		if line == "" {
			continue
		}
		if strings.HasPrefix(line, "@") {
			label := strings.Join(strings.Split(line, ""), "")
			label = label[:len(label)-1]
			if _, exists := labels[label]; !exists {
				labels[label] = len(instructions)
			}
		} else if !strings.HasPrefix(line, "$") {
			instructions = append(instructions, line)
		} else {
			vars = append(vars, line)
		}
	}
	return vars, labels, instructions
}

func validateInstructions(instructions []string, vars, labels []string) bool {
	hexChars := "0123456789abcdefABCDEF"
	binaryChars := "01"

	isValidAddress := func(address, opcode string) bool {
		if opcode == "set" || opcode == "clr" {
			validFlagNumbers := map[string]bool{
				"0": true, "1": true, "2": true, "3": true,
				"#0": true, "#1": true, "#2": true, "#3": true,
				"#d0": true, "#d1": true, "#d2": true, "#d3": true,
				"#h0": true, "#h1": true, "#h2": true, "#h3": true,
			}
			return validFlagNumbers[address]
		}

		if contains(vars, address) || contains(labels, address) {
			return true
		}

		if strings.HasPrefix(address, "#") {
			switch address[1] {
			case 'd':
				_, err := strconv.Atoi(address[2:])
				return err == nil
			case 'h':
				for _, c := range address[2:] {
					if !strings.ContainsRune(hexChars, c) {
						return false
					}
				}
				return true
			case 'b':
				for _, c := range address[2:] {
					if !strings.ContainsRune(binaryChars, c) {
						return false
					}
				}
				return true
			default:
				_, err := strconv.Atoi(address[1:])
				return err == nil
			}
		}

		_, err := strconv.Atoi(address)
		return err == nil
	}

	for _, instruction := range instructions {
		instructionParts := strings.Fields(instruction)
		if len(instructionParts) == 0 {
			fmt.Printf("[ ERROR : Invalid Instruction ] --> '%s'\n", instruction)
			return false
		}

		opcode := instructionParts[0]
		if !contains(OpCodes, opcode) {
			fmt.Printf("[ ERROR : Invalid Instruction ] --> '%s'\n", instruction)
			return false
		}

		if (contains(UnaryOpCodes, opcode) && len(instructionParts) > 1) ||
			(contains(BinaryOpCodes, opcode) && len(instructionParts) < 2) {
			fmt.Printf("[ ERROR : Invalid Instruction ] --> '%s'\n", instruction)
			return false
		}

		address := "0"
		if len(instructionParts) > 1 {
			address = instructionParts[1]
		}

		if !isValidAddress(address, opcode) {
			fmt.Printf("[ ERROR : Invalid Instruction ] --> '%s'\n", instruction)
			return false
		}
	}

	return true
}

func translate(sourceFile, outputFile string) {
	preprocessedData := preprocess(sourceFile)

	vars, labels, instructions := varsAndInstructions(preprocessedData)

	labelsList := keys(labels)
	varsDict := make(map[string]int)
	varsList := []string{}

	for _, v := range vars {
		parts := strings.Split(v, " = ")
		varName := parts[0]
		varValue := "0"
		if len(parts) > 1 {
			varValue = parts[1]
		}

		value := 0
		if strings.HasPrefix(varValue, "#") {
			base := map[byte]int{'d': 10, 'h': 16, 'b': 2}[varValue[1]]
			v, _ := strconv.ParseInt(varValue[2:], base, 0)
			value = int(v)
		} else {
			value, _ = strconv.Atoi(varValue)
		}

		varsDict[varName] = value
		if !contains(varsList, varName) {
			varsList = append(varsList, varName)
		}
	}

	if len(instructions) == 0 || instructions[len(instructions)-1] != "hlt" {
		instructions = append(instructions, "hlt")
	}

	if !validateInstructions(instructions, varsList, labelsList) {
		fmt.Println("Translation Failed!")
		return
	}

	linesWithResolvedAddresses := []string{}
	varInit := len(instructions)

	for _, instruction := range instructions {
		parts := strings.Fields(instruction)
		opcode := parts[0]
		var address string

		if len(parts) > 1 {
			address = parts[1]
		} else {
			address = "0"
		}

		if strings.HasPrefix(address, "#") {
			base := map[byte]int{'d': 10, 'h': 16, 'b': 2}[address[1]]
			v, _ := strconv.ParseInt(address[2:], base, 0)
			address = fmt.Sprintf("%d", v)
		}

		if contains(varsList, address) {
			address = fmt.Sprintf("%d", varInit+indexOf(varsList, address))
		} else if contains(labelsList, address) {
			address = fmt.Sprintf("%d", labels[address])
		}

		addrInt, _ := strconv.Atoi(address)
		opcodeHex := opcodeHexValue(opcode)
		linesWithResolvedAddresses = append(linesWithResolvedAddresses, opcodeHex+fmt.Sprintf("%05x", addrInt))
	}

	for _, varName := range varsList {
		linesWithResolvedAddresses = append(linesWithResolvedAddresses, fmt.Sprintf("%07x", varsDict[varName]))
	}

	ramContent := "v3.0 hex words addressed\n00000: " + strings.Join(linesWithResolvedAddresses, " ")

	file, err := os.Create(outputFile)
	if err != nil {
		fmt.Println("Error creating file:", err)
		os.Exit(1)
	}
	defer file.Close()

	file.WriteString(ramContent)
	fmt.Printf("Translation Successfully Done.\nOutput file: '%s'\n", outputFile)
}

func indexOf(slice []string, item string) int {
	for i, v := range slice {
		if v == item {
			return i
		}
	}
	return -1
}

func contains(slice []string, item string) bool {
	return indexOf(slice, item) != -1
}

func keys(m map[string]int) []string {
	keys := []string{}
	for k := range m {
		keys = append(keys, k)
	}
	return keys
}

func main() {
	if len(os.Args) != 3 {
		fmt.Println("Usage: go run translator.go <source_file> <output_file>")
		return
	}
	sourceFile := os.Args[1]
	outputFile := os.Args[2]
	translate(sourceFile, outputFile)
}
